yuyo/chunk_tracker.py

```diff
@@ -0,0 +1,596 @@
+    """
+
+    __slots__ = ("_app", "_data", "_shard")
+
+    def __init__(self, app: hikari.RESTAware, shard: hikari.api.GatewayShard, data: _RequestData, /) -> None:
+        """Initialise a chunk request finished event.
+
+        This should never be initialised directly.
+        """
+        self._app = app
+        self._data = data
+        self._shard = shard
+
+    @property
+    def app(self) -> hikari.RESTAware:
+        # <<inherited docstring from hikari.events.base_events.Event>>.
+        return self._app
+
+    @property
+    def shard(self) -> hikari.api.GatewayShard:
+        # <<inherited docstring from hikari.events.shard_events.ShardEvent>>.
+        return self._shard
+
+    @property
+    def chunk_count(self) -> int:
+        """The amount of chunk events which should've been received for this request."""
+        assert self._data.chunk_count
+        return self._data.chunk_count
+
+    @property
+    def first_received_at(self) -> datetime.datetime:
+        """When the first response was received."""
+        return self._data.first_received_at
+
+    @property
+    def guild_id(self) -> hikari.Snowflake:
+        """Id of the guild this chunk request was for."""
+        return self._data.guild_id
+
+    @property
+    def last_received_at(self) -> datetime.datetime:
+        """When the last response was received."""
+        return self._data.last_received_at
+
+    @property
+    def missed_chunks(self) -> typing.Collection[int]:
+        """Collection of the chunk responses which were missed (if any)."""
+        return self._data.missing_chunks or ()
+
+    @property
+    def not_found_ids(self) -> typing.Collection[hikari.Snowflake]:
+        """Collection of the User IDs which weren't found.
+
+        This is only relevant when `users` was specified while requesting the members.
+        """
+        return self._data.not_found_ids
+
+
+class ChunkingFinishedEvent(hikari.Event):  # TODO: better name?
+    """Event that's dispatched when the startup chunking has finished for the bot.
+
+    This indicates that any cache member and presences resources should be
+    complete globally.
+
+    This will only be fired after bot startups.
+    """
+
+    __slots__ = ("_app",)
+
+    def __init__(self, app: hikari.RESTAware, /) -> None:
+        """Initialise a chunking finished event.
+
+        This should never be initialised directly.
+        """
+        self._app = app
+
+    @property
+    def app(self) -> hikari.RESTAware:
+        # <<inherited docstring from hikari.events.base_events.Event>>.
+        return self._app
+
+
+class ShardChunkingFinishedEvent(hikari.ShardEvent):
+    """Event that's dispatched when the startup chunking has finished for a shard.
+
+    This indicates that any cache member and presences resources should be
+    complete for guilds covered by this shard.
+
+    This will be fired after every shard identify which triggers chunking
+    (including re-identifies).
```

Comment:
Rn it'll trigger this if literally none of the shards's guild were chunked, prob don't